rust/xaynet-analytics/src/controller.rs

```diff
@@ -148,3 +194,392 @@ impl AnalyticsController {
+        let controller_data = ControllerData::new(timestamp);
+        assert!(controller_data
+            .save(initial_controller.db(), CollectionNames::CONTROLLER_DATA)
+            .is_ok());
+
+        // init controller again, to read self.last_time_data_sent from db
+        assert!(initial_controller.dispose().is_ok());
+        let controller = get_controller(test_name, Some(12));
+        assert_eq!(controller.did_send_already_in_this_period(), true);
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_did_send_already_in_this_period_outside_12h() {
+        let test_name = "test_did_send_already_in_this_period_outside_12h";
+        let initial_controller = get_controller(test_name, Some(12));
+
+        let timestamp = Utc::now() - Duration::hours(13);
+        let controller_data = ControllerData::new(timestamp);
+        assert!(controller_data
+            .save(initial_controller.db(), CollectionNames::CONTROLLER_DATA)
+            .is_ok());
+
+        // init controller again, to read self.last_time_data_sent from db
+        assert!(initial_controller.dispose().is_ok());
+        let controller = get_controller(test_name, Some(12));
+        assert_eq!(controller.did_send_already_in_this_period(), false);
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_did_send_already_in_this_period_inside_6h() {
+        let test_name = "test_did_send_already_in_this_period_inside_6h";
+        let initial_controller = get_controller(test_name, Some(6));
+
+        let timestamp = Utc::now();
+        let controller_data = ControllerData::new(timestamp);
+        assert!(controller_data
+            .save(initial_controller.db(), CollectionNames::CONTROLLER_DATA)
+            .is_ok());
+
+        // init controller again, to read self.last_time_data_sent from db
+        assert!(initial_controller.dispose().is_ok());
+        let controller = get_controller(test_name, Some(6));
+        assert_eq!(controller.did_send_already_in_this_period(), true);
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_did_send_already_in_this_period_outside_6h() {
+        let test_name = "test_did_send_already_in_this_period_outside_6h";
+        let initial_controller = get_controller(test_name, Some(6));
+
+        let timestamp = Utc::now() - Duration::hours(7);
+        let controller_data = ControllerData::new(timestamp);
+        assert!(controller_data
+            .save(initial_controller.db(), CollectionNames::CONTROLLER_DATA)
+            .is_ok());
+
+        // init controller again, to read self.last_time_data_sent from db
+        assert!(initial_controller.dispose().is_ok());
+        let controller = get_controller(test_name, Some(6));
+        assert_eq!(controller.did_send_already_in_this_period(), false);
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_did_send_already_in_this_period_outside_twice_6h() {
```

Comment:
You mean to check twice within the same period if it has sent data? There's no test like that at the moment.

Not sure if it's clear, but the twice/thrice refers to 2 or 3 X hours away from `Utc::now()`.

Basically the only difference is that `outside_6h` test means the 7 below comes from 1*6 hours + 1:
`let timestamp = Utc::now() - Duration::hours(7);`

`outside_twice_6h` gives 2*6 hours + 1 = 13
`outside_thrice_6h` gives 3*6 hours + 1 = 19

Maybe i'm being obvious? Not sure it's clear enough. Happy to rename stuff to make it more clear

Reply:
thanks, yes I was indeed getting that confused. clear now.

Reply:
something else worth mentioning. taking the example of `outside_twice_6h`, what if the test runs at say, 11am? the `timestamp` is set at 10pm the _previous_ day - which is not even in the first period(!), since the first period always starts at 0000 of the _current_ day. your assert check
```rust
assert_eq!(!controller.did_send_already_in_this_period());
```
still passes all the same as if the test was run at say 2pm (in which case `timestamp` _does_ fall within the first period) so I guess this is fine. but these are like 2 different scenarios and it wasn't obvious that they're conflated in this (and other similar) test.

perhaps at a later date, we can split this up into separate test cases, and explicitly set the timestamps rather than have the nondeterminism of `Utc::now()`.